### ISSUE TITLE
Error when running ADAP chromatogram builder

### ISSUE THREAD

**Issue by mmzdouc** [mzmine/mzmine3#609]
My operating system: Ubuntu 20.04
My MZmine version: MzMine 3.0.11-beta

Dear developers,
while testing MzMine 3, I encountered a bug while trying to run the ADAP chromatogram builder algorithm on my data:

`2022-03-18 15:57:15 SEVERE io.github.mzmine.taskcontrol.impl.WorkerThread run Unhandled exception java.lang.IllegalArgumentException: Invalid range: [8.807934..8.783433] while processing task Detecting chromatograms in Qtof_P_E2_Wasteblank4.mzML
java.lang.IllegalArgumentException: Invalid range: [8.807934..8.783433]
	at com.google.common.collect.Range.<init>(Range.java:358)
	at com.google.common.collect.Range.create(Range.java:156)
	at com.google.common.collect.Range.closed(Range.java:181)
	at io.github.mzmine.datamodel.featuredata.FeatureDataUtils.getRtRange(FeatureDataUtils.java:73)
	at io.github.mzmine.datamodel.featuredata.FeatureDataUtils.recalculateIonSeriesDependingTypes(FeatureDataUtils.java:262)
	at io.github.mzmine.datamodel.featuredata.FeatureDataUtils.recalculateIonSeriesDependingTypes(FeatureDataUtils.java:244)
	at io.github.mzmine.datamodel.features.ModularFeature.<init>(ModularFeature.java:204)
	at io.github.mzmine.util.FeatureConvertors.ADAPChromatogramToModularFeature(FeatureConvertors.java:98)
	at io.github.mzmine.modules.dataprocessing.featdet_adapchromatogrambuilder.ModularADAPChromatogramBuilderTask.run(ModularADAPChromatogramBuilderTask.java:271)
	at io.github.mzmine.taskcontrol.impl.WorkerThread.run(WorkerThread.java:57)`

This happened with all three files that chromatogram building was run on. MzMine did not crash, but the algorithm froze and had to be cancelled.

For more information, please see the logfile attached.
[mzmine-log-error-18-03-22.log](https://github.com/mzmine/mzmine3/files/8305572/mzmine-log-error-18-03-22.log)

Please let me know if you need any more info from my side.

Kind regards,
Mitja Zdouc

EDIT: I fell back to MzMine2.53 and there, the ADAP chromatogram builder works and does not produce this problem.

**Comment by SteffenHeu** [comment]
Hi Mitja,

can you please upload the data file and post the parameters you used to produce this error?

**Comment by mmzdouc** [comment]
Hi Steffen,

there were three files, and since they were quite large, I will just upload one of them if that is alright - they all produced the same 
error:

[https://wetransfer.com/downloads/468db7183ca6a0364219dbbd7961ef3620220318194436/a032d2b1ae002a57bc8d5ae41180076220220318194611/f7266f](https://deref-gmx.net/mail/client/niRHhEqkOKI/dereferrer/?redirectUrl=https%3A%2F%2Fwetransfer.com%2Fdownloads%2F468db7183ca6a0364219dbbd7961ef3620220318194436%2Fa032d2b1ae002a57bc8d5ae41180076220220318194611%2Ff7266f) 

Parameters were set as follows: 

- Mass detection: noise level MS1: 5E1 (everything else default or empty)
- Mass detection: noise level MS2: 1E1 (everything else default or empty)
- ADAP chromatogram builder: min scans: 8; group int thresh: 1E2; min high int: 5E2; m/z acc: 10ppm;

At this point, the program stopped.

I hope that helps!

Kind regards,

Mitja Zdouc

**Comment by robinschmid** [comment]
Hey Mitja, 
There is something wrong with the scan numbers and sorting of the scans. You can double click on a file to see the data overview.
I guess you measured with lock mass? But somehow there are multiple scans with the same scan number. The lockmass scans follow one numbering, the MS1 scans another, and MS2 scans are numbered all over the place - with consecutive MS2 not even having consecutive numbers. 

It seems like each individual scan event starts with 1: MS1, MS1 lockmass, MS2 a, b, c.
Never seen anything like it.

1. What instrument is it?
2. How did you convert the files? (did you try the latest MS convert version?)
3. Can you apply lock mass calibration before converting to mzML and then leave out those scans?

![image](https://user-images.githubusercontent.com/10366914/159263101-086e9371-33d8-4896-9e47-62c360c553ab.png)
![image](https://user-images.githubusercontent.com/10366914/159262565-851f6c67-6cd5-4d20-8eed-06bb3a76cb30.png)

**Comment by mmzdouc** [comment]
Hey Robin, 

thanks for your suggestions! 

The data is coming from a Waters Synapt G2-Si hybrid quadrupole-traveling wave ion mobility (TWIM)-time-of-flight (TOF)
mass spectrometer. I received the data as part of a benchmarking project, so I don't know the exact procedure that was used to generate the files. I will write to the initiators and inform them about the unusual scan numbers and that it might have to do with lock mass calibration. Also, since I have access to the raw data as well, I might try to reconvert and/or read the raw data with MzMine3.

Still, it is interesting to note that MzMine2.53 was able to process the data, but not MzMine3.

**Comment by robinschmid** [comment]
We switched the mzML parser for MZmine 3. But still weird that MZmine 2 can process it. The scan numbers are weird. Maybe we applied sorting by retention time rather than scan number in MZmine 2, but I do not think so. 
The scan number like this will very likely screw up a few other modules as well that rely on scan numbers following the same sorting like retention time. 
Are the lock mass scan left out in MZmine 2?

**Comment by mmzdouc** [comment]
Hey Robin,

I checked the files again with MzMine2.53 and there, the scan numbers seem to be fine, they follow the expected numbering (ascending). I looked at the file that I provided you with (Qtof_P_E2_Wasteblank4.mzML), but the other files look the same.

![Screenshot from 2022-03-23 08-37-50](https://user-images.githubusercontent.com/95709447/159647473-74895abf-0dce-4649-8d16-e3491ff87460.png)

During import, there was only one warning: 
`[8:29:30|WARNING|uk.ac.ebi.jmzml.MzMLElement]: MzIdentML Configuration file: jar:file:/home/user/MZmine-2.53-Linux/lib/jmzml-1.7.11.jar!/defaultMzMLElement.cfg.xml`

Please find the log below:
[Log-MzMine2.53.log](https://github.com/mzmine/mzmine3/files/8330934/Log-MzMine2.53.log)

**Comment by robinschmid** [comment]
In MZmine 2, did you use the scan definition filter to show the MS1 scans without the lockmass scans?

It seems like MZmine 2 had a "feature" to renumber the scans automatically... but also because we needed that. The new data structure allows any scan numbers. 

I think it is also important to link to the correct scan numbers, there are other tools like the metabolomics USI resolver and the GNPS dashbord that rely on the scan numbers. They might also have a problem with your datafile.

**Comment by robinschmid** [comment]
Did you try the latest ms convert? Or what converter are you using?

**Comment by mmzdouc** [comment]
All the results were "vanilla" - I did not use the scan definition filter. 
I received the .mzML-files from a collaborator, so I have only limited information on the conversion. This is what they did, according to the documentation: 

> All .raw files were converted to .mzML using ProteoWizard MSConvert using default settings with the following exceptions:
binary encoding precision set to 32-bit, did not use zlib compression, the filter Peak Picking for vendor msLevel 1-2 added
as first step of processing.

Out of curiosity, I tried to convert again from the raw data using msconvert, but the behaviour persisted when opening the newly generated .mzML file with MzMine 3. 

I have informed the group that generated the LCMS data about the problem and that it seems to be linked to their data.
